Pysmurf controller tasks

jlashner opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
This PR implements many sodetlib operations as tasks in the pysmurf-controller agent. I tested most of these at UCSD, however we warmed up before I was able to test the `check_state` and `bias_dets` tasks. I'll either test this when we cool back down, or on another system if available before merging, but I think we can get the review out of the way before then.

## Description
<!--- Describe your changes in detail -->
Adds a bunch of tasks to the pysmurf-controller agent.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
So we can better use Smurf with OCS!

Resolves #252.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
Tested most functions at UCSD, but still need to test a couple of them.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] Unless I am preparing a release, I have opened this PR onto the `develop` branch.

mhasself commented on agents/pysmurf_controller/pysmurf_controller.py:
```diff
@@ -233,33 +257,336 @@ def abort(self, session, params=None):
+            session.set_status('running')
+            S, cfg = self._get_smurf_control(session=session)
+            bsa = bias_steps.take_bgmap(S, cfg, **params['kwargs'])
+            nchans_per_bg = [0 for _ in range(13)]
+            for bg in range(12):
+                nchans_per_bg[bg] = int(np.sum(bsa.bgmap == bg))
```
Define BIAS_COUNT = 12 somewhere and use that.

mhasself commented on agents/pysmurf_controller/pysmurf_controller.py:
```diff
@@ -233,33 +257,336 @@ def abort(self, session, params=None):
+            if not acquired:
+                return False, f"Operation failed: {self.lock.job} is running."
+
+            session.set_status('starting')
```
'running', right?

mhasself commented on agents/pysmurf_controller/pysmurf_controller.py:
```diff
@@ -233,33 +257,336 @@ def abort(self, session, params=None):
+        """bias_dets(rfrac=(0.3, 0.6), kwargs=None)
+
+        Biases detectors to a target Rfrac value or range. This function uses
+        IV results to determine voltages for each bias-group. If rfrac is set
```
Forgive my ignorance of sodetlib, but what IV is used?  (It looks like it doesn't acquire a new IV curve here, which is great... so does cfg know this somehow?)

jlashner replied:
By default, whenever an IV is taken the analyzed IV output is saved and it's path is written to the detconfig object as `iv_file`. So by default bias_dets uses the most recent IV, but can take an analyzed IV object as input to use as a keyword argument. It's a simple change to allow these functions to also optionally take a path to an analyzed IV file as a kwarg, but that would be a change in the sodetlib repo, and wouldn't require modification to this task since we're already passing arbitrary kwargs through.

mhasself commented on agents/pysmurf_controller/pysmurf_controller.py:
```diff
@@ -233,33 +257,336 @@ def abort(self, session, params=None):
+        """uxm_setup(bands=None, kwargs=None)
+
+        **Task** - Task to run first-time setup procedure for a UXM. This
+        will run the following operations:
```
Maybe mention how long each step takes?

mhasself commented on agents/pysmurf_controller/pysmurf_controller.py:
```diff
@@ -233,33 +257,336 @@ def abort(self, session, params=None):
+        """take_noise(duration=30., kwargs=None)
+
+        **Task** - Task to take a short timestream and calculate noise
+        statistics. Median white noise level for each band will be stored in
+        the session data. See the `sodetlib noise docs 
```
I don't think those noise levels are making it into session.data ...

BrianJKoopman replied:
I think it's happening here: https://github.com/simonsobs/sodetlib/blob/86bc88eaa00d7ba0c641077b5ccbe1e72682430b/sodetlib/noise.py#L537

The info stored in `session.data` should be added to this docstring in the `Notes:` section.

mhasself reviewed (changes requested):
Minor comments -- great to see these taking shape!

BrianJKoopman reviewed (changes requested):
Just a few small comments/questions.

Some other general items:
- Remove `agents/test_pysmurf_controller.py` to get rid of the import that's causing tests to fail.

- Docs are failing due to sodetlib, needs mocking or installation (see: https://readthedocs.org/projects/socs/builds/16624664/):
```
  File "../agents/pysmurf_controller/pysmurf_controller.py", line 10, in <module>
    import sodetlib as sdl
ModuleNotFoundError: No module named 'sodetlib'

Exception occurred:
  File "../agents/pysmurf_controller/pysmurf_controller.py", line 10, in <module>
    import sodetlib as sdl
ModuleNotFoundError: No module named 'sodetlib'
The full traceback has been saved in /tmp/sphinx-err-m0gn5wug.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
```

- Small wording thing, but I'd suggest documenting tasks like "**Task** - Check the current state of the smurf.", rather than "**Task** - Task to check the current state..." Similarly for Processes.

jlashner commented on agents/pysmurf_controller/pysmurf_controller.py:
```diff
@@ -233,33 +257,336 @@ def abort(self, session, params=None):
-    def tune_squids(self, session, params=None):
-        """tune_squids(args=[], log=True)
+    def check_state(self, session, params=None):
+        """check_state()
```
that makes sense to me!

jlashner commented on agents/pysmurf_controller/pysmurf_controller.py:
```diff
@@ -86,18 +93,22 @@ class PysmurfController:
 
+        self.slot = args.slot
+        if self.slot is None:
+            self.slot = os.environ['SLOT']
```
oops, it should be in the docker-compose entry. I'll update the docs.

jlashner commented:
Alright! I think I hit all the suggestions. Added examples for session.data, made check_state into a process (and tested for robustness against smurf-streamer dropouts), and I think I fixed the tests / docs. Let me know if there's anything else I should do before merging!

BrianJKoopman reviewed (changes requested):
Changes related to my comments look good. I noticed a docstring that needs updating.

Other than that, could you merge in the latest 'develop'? That'll catch up to recent updates to the github actions workflows, and there are some conflicts in `docs/agents/pysmurf/pysmurf-controller.rst` to resolve.

BrianJKoopman commented on agents/pysmurf_controller/pysmurf_controller.py:
```diff
@@ -307,7 +307,8 @@ def check_state(self, session, params=None):
                     stream_tag = reg.stream_tag.get(**kw),
-                    last_update = time.time()
+                    last_update = time.time(),
+                    stream_id = cfg.stream_id,
```
Should be added to the `session.data` example in the docstring.

jlashner replied:
Good catch, thanks! Just added.

BrianJKoopman reviewed (changes requested):
Another small docstring addition related to additional changes in the last commit.

BrianJKoopman reviewed (approved):
Great, looks good to me!

I haven't really looked over any changes related to suggestions from @mhasself though.
